add math_epsilon for number diffs that are close

flowolf opened the pull request:
Hi!

thanks for an awesome library!

I've been using a patch for a few months now, that I wanted to send upstream, as I think others might find it useful.

**Use case**:
For some sensor data derived and computed values must lie in a certain range. It does not matter that they are off by e.g. 1e-5.

To check against that the `math` core module provides the valuable `isclose()` function. It evaluates the being close of two numbers to each other, with reference to an epsilon (abs_tol). This is superior to the format function, as it evaluates the mathematical representation and not the string representation.

**New introduced feature**:
My current implementation works well to compare numbers in objects. As the number is used by `math`, it cannot currently handle the hashing of values, which is done when `ignore_order` is used. Other than that it is solid, as only `math.isclose()` is called.

~~I introduce the `math_close` boolean parameter. If set to True it will use `math.isclose()` to evaluate numbers. The new optional `epsilon` takes a tolerance value which is passed to `math.isclose()`. Any numbers that are within the tolerance will not report as being different. Any numbers outside of that tolerance will show up as different.~~

I introduce the `math_epsilon` parameter. If set will use `math.isclose()` to evaluate numbers. It defines a tolerance value which is passed to `math.isclose()`. Any numbers that are within the tolerance will not report as being different. Any numbers outside of that tolerance will show up as different.

I also added tests for this functionality.

seperman commented:
Hi @flowolf 
Thanks for the PR! Yep the reason we didn't end up using Math.isClose was exactly the hashing issue.

1. If you have time, can you please make it so it logs some warning that isClose will be ignored when ignore_order = True?

2. Also do you think we need both `math_close` and `epsilon` as parameters? Can it be just something like `math_close_epsilon` which has a default of None but when it has any other positive numeric value, it enables this feature? The reason I'm asking that is we already have many parameters on DeepDiff and it will be better if we can reduce the number of parameters we will be adding.

Thanks again!

flowolf commented:
@seperman Thanks for the review!

I added a warning and changed the parameters.
the new parameter is now `math_epsilon` (a little shorter). if set to None (default) it's not used. if set it will be passed to `math.isclose()` and enable the math comparison functionality.

adapted the tests accordingly.

seperman commented:
Hi @flowolf 
Awesome. Thank you for the PR! I need to add some more tests to the dev branch and will release it probably in the following week.

seperman commented:
Hi @flowolf 
Just wanted to keep you posted that DeepDiff 5.2.1 is released and it includes your PR.
Thanks for contributing to DeepDiff!!

flowolf commented:
thanks that's great news.

martin-ueding commented:
[The documentation](https://deepdiff.readthedocs.io/en/latest/diff.html) is still for version 4.3.0 and doesn't include this option. This way I didn't see it at first.

seperman commented:
I moved the docs from read the docs to my own website mid summer 2020. The link on github also points to the new location.

https://zepworks.com/deepdiff/current/

Sep Dehpour

> On Jan 18, 2021, at 9:53 AM, Martin Ueding <notifications@github.com> wrote:
> 
> ﻿
> The documentation is still for version 4.3.0 and doesn't include this option. This way I didn't see it at first.
> 
> —
> You are receiving this because you modified the open/close state.
> Reply to this email directly, view it on GitHub, or unsubscribe.

martin-ueding commented:
Maybe it would be a good idea to delete the old version on Read The Docs completely as it is still high in the search results for “deepdiff”.

seperman commented:
Good point. Ok the docs on Read the docs are deleted.
